netlify/netlify-plugin-nextjs
# fix: prepend basePath to static file URLs

### Summary

This partially fixes a complex bug that could manifest in several ways. There is an additional internal bugfix that needs to land in the proxy before this can be finally fixed.

#### Some of the details of the issue:
- When Next attempts to read a pre-rendered page file from disk that has been moved to the CDN, the function makes a request to load that file, which it then caches. It constructs that URL by appending the requested path to the site domain. e.g. if Next tries to read `/page.html` and that file has been moved to the CDN, it will try to load `https://example.com/page.html`. It finds this domain using `event.rawUrl`.
- Previously, this wasn't taking account of `basePath`, so was loading the URL without the basePath. This was fine when loading the site directly, because the actual files were in that place too.
- There is a bug in proxy that means that the `rawUrl` passed to the proxied function is the original URL, rather than the proxied URL. There is no way for the function to know its own URL.
- Because of this, when proxying from `https://originsite/basePath` to `https://targetsite/basePath`, it would try to load the files from `originsite` instead of `targetsite`, and because the path didn't match the proxy pattern, it was failing to load, giving a 404
- Because of the way the Next server is memoized and files are cached, if the first request for a page was directly to the site rather than via the proxy, then it would correctly download and cache the file, and it would be available for subsequent requests via the proxy. This is why it was so hard to debug.

#### This fix
This fix ensures that the files moved to the CDN are moved under the correct basePath, meaning that the file `docs.html` is moved to `basePath/docs.html`. The proxy bug is still in existence, but because the basePath matches, the CDN request will also be proxied.

This PR also adds a test with with a basePath set. There aren't currently any cypress tests for it, but it can be manually tested. The proxy site is pointing at the deploy preview for this PR. 

### Test plan

1. Visit [the proxy site](https://nf-test-gatsby-proxy.netlify.app/)
2. Click through to the Next site, then fallback and dynamic pages
3. Ensure that they are not 404
 
### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal

Fixes #965 

![capybara and rabbits](https://pbs.twimg.com/media/FLGPpAqWQAYAEnB?format=jpg&name=small)

### Standard checks:




---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.
